File: app/src/main/java/lee/jaebeom/WikiPortal/java/SettingFragment.java
/*
package lee.jaebeom.WikiPortal.setting;

import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.v7.preference.CheckBoxPreference;
import android.support.v7.preference.ListPreference;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceCategory;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.preference.PreferenceManager;
import android.util.Log;

import java.util.Map;

import lee.jaebeom.WikiPortal.R;
import lee.jaebeom.WikiPortal.util.SharedPreference;
import lee.jaebeom.WikiPortal.wiki.WikiData;
import lee.jaebeom.WikiPortal.wiki.WikiModel;

public class SettingFragment extends PreferenceFragmentCompat implements SettingContract.View, SharedPreferences.OnSharedPreferenceChangeListener{

    SettingContract.Presenter presenter;
    ListPreference listPreference;
    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.pref_setting);

        PreferenceCategory category = (PreferenceCategory)findPreference("setting_title");
        for (Map.Entry<String, WikiModel> entry: WikiData.wikiURL.entrySet()){
            CheckBoxPreference checkBoxPreference = new CheckBoxPreference(this.getActivity());
            checkBoxPreference.setTitle(entry.getKey());
            checkBoxPreference.setKey(entry.getKey());
            checkBoxPreference.setChecked(false);
            category.addPreference(checkBoxPreference);
        }

        listPreference = (ListPreference)findPreference("wikipedia_language_setting");
        listPreference.setSummary(
                listPreference.getEntries()[
                    listPreference.findIndexOfValue(
                            SharedPreference.getStringPreferences(getContext(), "wikipedia_language_setting")
                    )
                ].toString());
        listPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                ListPreference listPreference = (ListPreference)preference;
                String langKey = listPreference.getEntries()[listPreference.findIndexOfValue(newValue.toString())].toString();  //ex) English, 한국어
                String langValue = newValue.toString();    //ex) en, ko

                listPreference.setSummary(langKey);
                WikiData.wikiURL.get(WikiData.Wikipedia).setLanguage(langValue);
                WikiData.wikiURL.get(WikiData.Wikipedia).setChangeURLLanguage(langValue);
                Log.d("listPreference", newValue.toString());
                return true;
            }
        });
    }

    @Override
    public void setDivider(Drawable divider) {
        super.setDivider(new ColorDrawable(Color.TRANSPARENT));
    }

    @Override
    public void setDividerHeight(int height) {
        super.setDividerHeight(0);
    }

    @Override
    public void _presenter(SettingContract.Presenter presenter) {
        this.presenter = presenter;
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getContext());
//        presenter.setUsed(sharedPreferences, key);
        Log.d("pref key", key);
        if (!key.equals("wikipedia_language_setting")){
            WikiData.wikiURL.get(key).setUse(sharedPreferences.getBoolean(key, false));
        }else{

        }
//        if (WikiData.wikiURL.get(key).isUse()){
//            WikiData.wikis++;
//        }else{
//            WikiData.wikis--;
//        }

    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);

    }
}
*/

File: app/src/main/java/lee/jaebeom/WikiPortal/setting/SettingFragment.kt
package lee.jaebeom.WikiPortal.setting

import android.content.SharedPreferences
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.v7.preference.*
import android.util.Log
import lee.jaebeom.WikiPortal.R
import lee.jaebeom.WikiPortal.util.SharedPreference
import lee.jaebeom.WikiPortal.wiki.WikiData
import lee.jaebeom.WikiPortal.wiki.WikiModel

/**
 * Created by leejaebeom on 2017. 11. 7..
 */
class SettingFragment : PreferenceFragmentCompat(), SettingContract.View, SharedPreferences.OnSharedPreferenceChangeListener {
    lateinit var presenter: SettingContract.Presenter
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.pref_setting)
        val category = findPreference("setting_title") as PreferenceCategory
        for (entry: Map.Entry<String, WikiModel> in WikiData.wikiURL.entries){
            val chechBox = CheckBoxPreference(this.activity)
            chechBox.title = entry.key
            chechBox.key = entry.key
            chechBox.isChecked = false
            chechBox.onPreferenceClickListener = object : Preference.OnPreferenceClickListener{
                override fun onPreferenceClick(preference: Preference?): Boolean {
                    val checkBox: CheckBoxPreference = preference as CheckBoxPreference
                    if (chechBox.isChecked){
                        entry.value.sequence = WikiData.useWikis.size + 1
                        SharedPreference.savePreference("${SharedPreference.keySequence}_${entry.key}", entry.value.sequence, context)
                        WikiData.useWikis.add(entry.value)
                    }else{
                        entry.value.sequence = 0
                        SharedPreference.savePreference("${SharedPreference.keySequence}_${entry.key}", entry.value.sequence, context)
                        WikiData.useWikis.remove(entry.value)   //사용하는 위키에서 제거
                        WikiData.orderSequence(context)   //순서 값 재설정
                    }
                    return true
                }
            }
            category.addPreference(chechBox)
        }

        val listPreference  = findPreference("wikipedia_language_setting") as ListPreference
        listPreference.summary = listPreference.entries[
                listPreference.findIndexOfValue(
                        SharedPreference.getStringPreferences(context, "wikipedia_language_setting")
                )].toString()

        listPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { preference, newValue ->
            var listPreference = preference as ListPreference
            var langKey = listPreference.entries[listPreference.findIndexOfValue(newValue.toString())].toString()
            var langValue = newValue.toString()

            listPreference.summary = langKey
            WikiData.wikiURL.get(WikiData.Wikipedia)?.language = langValue
            WikiData.wikiURL.get(WikiData.Wikipedia)?.setChangeURLLanguage(langValue)
            Log.d("listPreference", newValue.toString())

            true
        }

        val wikiSequencePreference = findPreference("wiki_sequence")


    }

    override fun setDivider(divider: Drawable?) {
        super.setDivider(ColorDrawable(Color.TRANSPARENT))
    }

    override fun setDividerHeight(height: Int) {
        super.setDividerHeight(0)
    }
    override fun _presenter(presenter: SettingContract.Presenter) {
        this.presenter = presenter
    }

    override fun onSharedPreferenceChanged(p0: SharedPreferences?, p1: String?) {
        if (!p1.equals("wikipedia_language_setting")){
            WikiData.wikiURL.get(p1)?.isUse = p0!!.getBoolean(p1, false)
        }
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }
}